### PR TITLE
feat: Add retry_if_exception_cause_type

### PR DESCRIPTION
**Pull request by Greesb**
Add a new retry_base class called `retry_if_exception_cause_type` that checks that the cause of the raised exception is of a certain type.

### CODE REVIEW

**Review by sileht — DISMISSED** [review]

**Review by JulesVautier — COMMENTED** [review]
Do we add a releasenote if the change doesn't impact customers ?

**Review by JulesVautier — APPROVED** [review]

File: tenacity/retry.py [hunk 1]
```diff
@@ -117,6 +117,29 @@ def __call__(self, retry_state: "RetryCallState") -> bool:
         return self.predicate(retry_state.outcome.exception())
 
 
+class retry_if_exception_cause_type(retry_base):
+    """Retries if the cause of the raised exception is of one or mor types."""
```
Comment (asqui) [inline tenacity/retry.py:121]: Typo in “…one or mor types”

**Review by asqui — COMMENTED** [review]
Is this change really necessary?

Why not just use `retry_if_exception(predicate)` with a lambda to achieve the same effect?

**Comment by Greesb** [comment]
> Is this change really necessary?
> 
> Why not just use `retry_if_exception(predicate)` with a lambda to achieve the same effect?

By "this change" do you mean the way it is implemented ? Or do you mean the new class on its own ?

If you're talking about the new class on its own, then a lot of existing `retry_base` class can be achieved with a predicate with lambda, it's just a quality of life improvement to already have a "native" feature that does it to not have to implement a lambda that does it.

If you're talking about the way it is implemented, then yeah it could also probably be done with a predicate, but it'll probably look really ugly. I just thought it looked better like this.

**Comment by asqui** [comment]
Sorry for not being clear, and thank you for your contribution.

I meant "why do we need this class", because I didn't see a big difference between:
```
retry_if_exception(lambda e: isinstance(e.cause, SomeException))
```
and
```
retry_if_exception_cause_type(SomeExcetion)
```
However, on closer inspection, I see that this is slightly different because your implementation considers all chained exceptions and not just the first.

I guess this is quite deliberate so it would be good to make this more clear in the docstring/release note -- my initial feedback was based on the docstring/release note from which I assumed it to be equivalent to the first snippet above.

(I don't think that "exception cause" makes it immediately obvious that this includes all causes up the chain of nested exceptions; at least not to me.)

As for the implementation, I don't think using `retry_if_exception` and a predicate will add or take away much (would remove the if statement and the `return False` but add a `def inner():` and `super()` in its place, no big deal either way) though I do see that all of the other `retry_*exception_*` classes follow the pattern of inheriting from `retry_if_exception` so this would be the first deviation from that...

**Comment by Greesb** [comment]
> Sorry for not being clear, and thank you for your contribution.
> 
> I meant "why do we need this class", because I didn't see a big difference between:
> 
> ```
> retry_if_exception(lambda e: isinstance(e.cause, SomeException))
> ```
> 
> and
> 
> ```
> retry_if_exception_cause_type(SomeExcetion)
> ```
> 
> However, on closer inspection, I see that this is slightly different because your implementation considers all chained exceptions and not just the first.
> 
> I guess this is quite deliberate so it would be good to make this more clear in the docstring/release note -- my initial feedback was based on the docstring/release note from which I assumed it to be equivalent to the first snippet above.
> 
> (I don't think that "exception cause" makes it immediately obvious that this includes all causes up the chain of nested exceptions; at least not to me.)
> 
> As for the implementation, I don't think using `retry_if_exception` and a predicate will add or take away much (would remove the if statement and the `return False` but add a `def inner():` and `super()` in its place, no big deal either way) though I do see that all of the other `retry_*exception_*` classes follow the pattern of inheriting from `retry_if_exception` so this would be the first deviation from that...

You're right the documentation was a bit unclear on what it was doing, i updated the docstring and the release note.
Let me know if it's better and/or if you've got any other suggestions.

**Comment by asqui** [comment]
Nice, that’s a lot clearer. No further feedback from me.

**Comment by AIGeneratedUsername** [comment]
Could you please update https://tenacity.readthedocs.io/en/latest/changelog.html alongside with new releases? The "8.1.0" is the latest at Pypi, but I see no changelog changes. Thanks.

**Pull request merged**